providers/common/ciphers/aes_gcm_s390x.inc

```diff
@@ -0,0 +1,325 @@
+        memcpy(&kma->j0, iv, ivlen);
+        kma->j0.w[3] = 1;
+        kma->cv.w = 1;
+    } else {
+        unsigned long long ivbits = ivlen << 3;
+        size_t len = S390X_gcm_ivpadlen(ivlen);
+        unsigned char iv_zero_pad[AES_GCM_IV_MAX_SIZE + S390X_IV_PAD_LEN];
+        /*
+         * The IV length needs to be zero padded to be a multiple of 16 bytes
+         * followed by 8 bytes of zeros and 8 bytes for the IV length.
+         * The GHASH of this value can then be calculated.
+         */
+        memcpy(iv_zero_pad, iv, ivlen);
+        memset(iv_zero_pad + ivlen, 0, len - ivlen);
+        memcpy(iv_zero_pad + len - sizeof(ivbits), &ivbits, sizeof(ivbits));
+        /*
+         * Calculate the ghash of the iv - the result is stored into the tag
+         * param.
+         */
+        s390x_kma(iv_zero_pad, len, NULL, 0, NULL, ctx->plat.s390x.fc, kma);
+        ctx->plat.s390x.fc |= S390X_KMA_HS; /* The hash subkey is set */
+
+        /* Copy the 128 bit GHASH result into J0 and clear the tag */
+        kma->j0.g[0] = kma->t.g[0];
+        kma->j0.g[1] = kma->t.g[1];
+        kma->t.g[0] = 0;
+        kma->t.g[1] = 0;
+        /* Set the 32 bit counter */
+        kma->cv.w = kma->j0.w[3];
+    }
+    return 1;
+}
+
+static int s390x_aes_gcm_cipher_final(PROV_AES_GCM_KEY *ctx, unsigned char *tag)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+    unsigned char out[S390X_OUT_LEN];
+
+    kma->taadl <<= 3;
+    kma->tpcl <<= 3;
+    s390x_kma(ctx->plat.s390x.ares, ctx->plat.s390x.areslen,
+              ctx->plat.s390x.mres, ctx->plat.s390x.mreslen, out,
+              ctx->plat.s390x.fc | S390X_KMA_LAAD | S390X_KMA_LPC, kma);
+
+    /* gctx->mres already returned to the caller */
+    OPENSSL_cleanse(out, ctx->plat.s390x.mreslen);
+
+    if (ctx->enc) {
+        ctx->taglen = AES_GCM_TAG_MAX_SIZE;
+        memcpy(tag, kma->t.b, ctx->taglen);
+    } else {
+        if (ctx->taglen < 0)
+            return 0;
+
+        if (CRYPTO_memcmp(tag, kma->t.b, ctx->taglen) != 0)
+            return 0;
+    }
+    return 1;
+}
+
+/*
+ * in_len = length of PlainText & CipherText
+ * For encryption:
+ *   in = Plaintext
+ *   out = CipherText + tag
+ * For decryption:
+ *   in = CipherText + Tag
+ *   out = returned plaintext
+ * tag is the calculated Authentication Tag (with a size of 16 bytes)
+ */
+static int s390x_aes_gcm_one_shot(PROV_AES_GCM_KEY *ctx,
+                                  unsigned char *aad, size_t aad_len,
+                                  const unsigned char *in, size_t in_len,
+                                  unsigned char *out,
+                                  unsigned char *tag, size_t taglen)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+
+    kma->taadl = aad_len << 3;
+    kma->tpcl = in_len << 3;
+    s390x_kma(aad, aad_len, in, in_len, out,
+              ctx->plat.s390x.fc | S390X_KMA_LAAD | S390X_KMA_LPC, kma);
+    memcpy(tag, kma->t.b, taglen);
+    return 1;
```

Comment:
Seems like i missed an sslcorrupt test failure last time. However, this function obviously has to compare tags instead of copying (and return 0 if unsuccsessful) in case of decryption.